lib/concurrent/struct/immutable_struct.rb

```diff
@@ -0,0 +1,87 @@
+require 'concurrent/struct/abstract_struct'
+require 'concurrent/synchronization'
+
+module Concurrent
+
+  # A thread-safe, immutable variation of Ruby's standard `Struct`.
+  #
+  # @see http://ruby-doc.org/core-2.2.0/Struct.html Ruby standard library `Struct`
+  module ImmutableStruct
+    include AbstractStruct
+
+    # @!visibility private
+    def initialize(*values)
+      ns_initialize(*values)
+    end
```

Comment:
That sounds like I've written it badly. It should not force calling to `synchronize` in cases like this. The default call to `ns_initialize` should be probably removed and left on `Synchronization::Object` users. I would like to avoid moving `ensure_ivar_visibility!` to class because it would create requirement to be implemented on class level which may create complications across different implementations. If you agree I'll remove the auto `ns_initialize` call replacing it with a helper module or a subclass.